File: Attributes/DistinctFieldAttribute.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

''' --------------------------------------------------------------------------------
''' <summary>
''' Specifies the field name that uniquely identifies each object 
''' within the collection. Typically, this is the field name of an identity or auto 
''' increment field. If the SubSetAttribute has been specified
''' then the strDistinctFieldName need only be unique within the subset not the 
''' entire table. The strDistinctFieldName and can be identical to the field name
''' specified with a KeyField attribute.
''' This attribute must be specified on a DatabaseObjects*UsingAttributes class.
''' This attribute is used to implement the IDatabaseObjects.DistinctFieldName
''' and IDatabaseObjects.DistinctFieldAutoIncrements functions.
''' </summary>
''' <example>
''' <code>
'''    &lt;DistinctField("CustomerID", bAutoIncrements:=True)&gt;
'''    Public Class Customers
'''        ...
''' </code>
''' </example>
''' --------------------------------------------------------------------------------
<AttributeUsage(AttributeTargets.Class, AllowMultiple:=False, Inherited:=True)> _
Public Class DistinctFieldAttribute
    Inherits Attribute

    Private pstrDistinctFieldName As String
    Private peFieldValueAutomaticAssignment As SQL.FieldValueAutoAssignmentType = SQL.FieldValueAutoAssignmentType.None

    ''' <summary>
    ''' Specifies the field name that uniquely identifies each object 
    ''' within the collection.
    ''' </summary>
    Public Sub New(ByVal strDistinctFieldName As String)

        Me.New(strDistinctFieldName, False)

    End Sub

    ''' <summary>
    ''' Specifies the field name that uniquely identifies each object 
    ''' within the collection. Typically, this is the field name of an identity or auto 
    ''' increment field in which case the bAutoIncrements value should be set to true.
    ''' </summary>
    Public Sub New(ByVal strDistinctFieldName As String, ByVal bAutoIncrements As Boolean)

        If strDistinctFieldName = String.Empty Then
            Throw New ArgumentNullException
        End If

        pstrDistinctFieldName = strDistinctFieldName
        If bAutoIncrements Then
            peFieldValueAutomaticAssignment = SQL.FieldValueAutoAssignmentType.AutoIncrement
        End If

    End Sub

    ''' <summary>
    ''' Specifies the field name that uniquely identifies each object 
    ''' within the collection. Typically, this is the field name of an identity or auto 
    ''' increment field in which case the bAutoIncrements value should be set to true.
    ''' </summary>
    Public Sub New(ByVal strDistinctFieldName As String, ByVal eAutomaticAssignment As SQL.FieldValueAutoAssignmentType)

        If strDistinctFieldName = String.Empty Then
            Throw New ArgumentNullException
        End If

        pstrDistinctFieldName = strDistinctFieldName
        peFieldValueAutomaticAssignment = eAutomaticAssignment

    End Sub

    Public ReadOnly Property Name() As String
        Get

            Return pstrDistinctFieldName

        End Get
    End Property

    Public ReadOnly Property AutomaticAssignment() As SQL.FieldValueAutoAssignmentType
        Get

            Return peFieldValueAutomaticAssignment

        End Get
    End Property

End Class

File: Attributes/DistinctFieldAttribute.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;

namespace DatabaseObjects
{
	/// --------------------------------------------------------------------------------
	/// <summary>
	/// Specifies the field name that uniquely identifies each object
	/// within the collection. Typically, this is the field name of an identity or auto
	/// increment field. If the SubSetAttribute has been specified
	/// then the strDistinctFieldName need only be unique within the subset not the
	/// entire table. The strDistinctFieldName and can be identical to the field name
	/// specified with a KeyField attribute.
	/// This attribute must be specified on a DatabaseObjects*UsingAttributes class.
	/// This attribute is used to implement the IDatabaseObjects.DistinctFieldName
	/// and IDatabaseObjects.DistinctFieldAutoIncrements functions.
	/// </summary>
	/// <example>
	/// <code>
	///    &lt;DistinctField("CustomerID", bAutoIncrements:=True)&gt;
	///    Public Class Customers
	///        ...
	/// </code>
	/// </example>
	/// --------------------------------------------------------------------------------
	[AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = true)]
	public class DistinctFieldAttribute : Attribute
	{
		private string pstrDistinctFieldName;
		private SQL.FieldValueAutoAssignmentType peFieldValueAutomaticAssignment = SQL.FieldValueAutoAssignmentType.None;
		
		/// <summary>
		/// Specifies the field name that uniquely identifies each object
		/// within the collection.
		/// </summary>
		public DistinctFieldAttribute(string strDistinctFieldName)
            : this(strDistinctFieldName, bAutoIncrements: false)
		{
		}
		
		/// <summary>
		/// Specifies the field name that uniquely identifies each object
		/// within the collection. Typically, this is the field name of an identity or auto
		/// increment field in which case the bAutoIncrements value should be set to true.
		/// </summary>
		public DistinctFieldAttribute(string strDistinctFieldName, bool bAutoIncrements)
		{
			if (String.IsNullOrEmpty(strDistinctFieldName))
				throw new ArgumentNullException();
			
			pstrDistinctFieldName = strDistinctFieldName;

			if (bAutoIncrements)
				peFieldValueAutomaticAssignment = SQL.FieldValueAutoAssignmentType.AutoIncrement;
		}
		
		/// <summary>
		/// Specifies the field name that uniquely identifies each object
		/// within the collection. Typically, this is the field name of an identity or auto
		/// increment field in which case the bAutoIncrements value should be set to true.
		/// </summary>
		public DistinctFieldAttribute(string strDistinctFieldName, SQL.FieldValueAutoAssignmentType eAutomaticAssignment)
		{
			if (String.IsNullOrEmpty(strDistinctFieldName))
				throw new ArgumentNullException();
			
			pstrDistinctFieldName = strDistinctFieldName;
			peFieldValueAutomaticAssignment = eAutomaticAssignment;
		}
		
		public string Name
		{
			get
			{
				return pstrDistinctFieldName;
			}
		}
		
		public SQL.FieldValueAutoAssignmentType AutomaticAssignment
		{
			get
			{
				return peFieldValueAutomaticAssignment;
			}
		}
	}
	
}
